browser/lib/util/crypto.js

```diff
@@ -98,84 +115,102 @@ var Crypto = (function() {
 
 	/**
-	 * Obtain a default CipherParams. This uses default algorithm, mode and
-	 * padding. If a key is specified this is used; otherwise a new key is generated
-	 * for the default key length. An IV is generated using the default
-	 * system SecureRandom.
-	 * A generated key may be obtained from the returned CipherParams
-	 * for out-of-band distribution to other clients.
-	 * @param key (optional) ArrayBuffer, Array, WordArray or base64 string, containing key
+	 * Obtain a complete CipherParams instance from the provided params, filling
+	 * in any not provided with default values, calculating a keyLength from
+	 * the supplied key, and validating the result.
+	 * @param params an object containing at a minimum a `key` key with value the
+	 * key, as either a binary (ArrayBuffer, Array, WordArray) or a
+	 * base64-encoded string. May optionally also contain: algorithm (defaults to
+	 * AES), mode (defaults to 'cbc')
 	 * @param callback (err, params)
 	 */
-	Crypto.getDefaultParams = function(key, callback) {
-		if(arguments.length == 1 && typeof(key) == 'function') {
-			callback = key;
-			key = undefined;
-		}
-		if(!key) {
-			generateRandom(DEFAULT_KEYLENGTH / 8, function(err, buf) {
-				if(err) {
-					callback(err);
-					return;
-				}
-				Crypto.getDefaultParams(buf, callback);
-			});
+	Crypto.getDefaultParams = function(params) {
+		var key;
+		/* Backward compatibility */
+		if((typeof(params) === 'function') || (typeof(params) === 'string')) {
+			Logger.deprecated('Crypto.getDefaultParams(key, callback)', 'Crypto.getDefaultParams({key: key})');
+			if(typeof(params) === 'function') {
+				Crypto.generateRandomKey(function(key) {
+					params(null, Crypto.getDefaultParams({key: key}));
+				})
+			} else {
+				callback = arguments[1];
+				callback(null, Crypto.getDefaultParams({key: params}));
+			}
 			return;
 		}
-		if (typeof(key) === 'string')
-			key = CryptoJS.enc.Hex.parse(key);
-		else
-			key = BufferUtils.toWordArray(key);   // Expect key to be an Array, ArrayBuffer, or WordArray at this point
-
-		var params = new CipherParams();
-		params.algorithm = DEFAULT_ALGORITHM;
-		params.key = key;
-		params.keyLength = key.words.length * (4 * 8);
-		params.mode = DEFAULT_MODE;
-		generateRandom(DEFAULT_BLOCKLENGTH, function(err, buf) {
-			params.iv = buf;
-			callback(null, params);
-		});
+
+		if(!params.key) {
+			throw new Error('Crypto.getDefaultParams: a key is required');
+		}
+
+		if (typeof(params.key) === 'string') {
+			key = CryptoJS.enc.Base64.parse(normaliseBase64(params.key));
+		} else {
+			key = BufferUtils.toWordArray(params.key); // Expect key to be an Array, ArrayBuffer, or WordArray at this point
+		}
+
+		var cipherParams = new CipherParams();
+		cipherParams.key = key;
+		cipherParams.algorithm = params.algorithm || DEFAULT_ALGORITHM;
+		cipherParams.keyLength = key.words.length * (4 * 8);
+		cipherParams.mode = params.mode || DEFAULT_MODE;
+		validateCipherParams(cipherParams);
```

Comment:
What happens if a keyLength is provided explicitly?  I assume it's just ignored as the keyLength is derived from the key. In Ruby I've checked for inconsistencies and indicated an error. Not sure it's necessary, but just a comment